docs/decisions/20240116-use-zio-failures-and-defects-effectively.md

```diff
@@ -231,6 +221,296 @@ Management Best Practices documentation.
+    val cxnIO =
+      sql"""
+           | INSERT INTO public.connection_records(
+           |   id,
+           |   created_at,
+           |   ...
+           | ) values (
+           |   ${record.id},
+           |   ${record.createdAt},
+           |   ...
+           | )
+        """.stripMargin.update
+
+    cxnIO.run
+      .transactWallet(xa)
+      .ensureOneAffectedRowOrDie
+  }
+}
+
+extension[Int](ma: RIO[WalletAccessContext, Int]) {
+  def ensureOneAffectedRowOrDie: URIO[WalletAccessContext, Unit] = ma.flatMap {
+    case 1 => ZIO.unit
+    case count => ZIO.fail(RuntimeException(s"Unexpected affected row count: $count"))
+  }.orDie
+}
+```
+
+#### Do not reflexively log errors
+
+The upper layer will automatically do so in an appropriate and consistent way using Tapir interceptor customisation.
+
+### Service Layer
+
+#### Reporting `404 Not Found` to user
+
+How can a service appropriately report a `404 Not Found` to a user that e.g. tries to update a record
+that does not exist in the database? Following the above rules, the `update` method will throw a defect that will be
+caught at the upper level and return a generic `500 Internal Server Error` to the user.
+
+For those cases where a specific error like `404` should be returned, it is up to the service to first call `find()`
+before `update()` and construct a `NotFound` failure, propagated through error channel, if it gives
+a `None` ([example](https://github.com/hyperledger-labs/open-enterprise-agent/blob/eb898e068f768507d6979a5d9bab35ef7ad4a045/connect/lib/core/src/main/scala/io/iohk/atala/connect/core/service/ConnectionServiceImpl.scala#L149)).
+
+Relying on the service layer to implement it will guarantee consistent behaviour regardless of the underlying database
+type (could be different RDMS flavour, No-SQL, etc.).
+
+```scala
+class ConnectionServiceImpl() extends ConnectionService {
+  override def markConnectionRequestSent(recordId: UUID):
+  ZIO[WalletAccessContext, RecordIdNotFound | InvalidStateForOperation, ConnectionRecord] =
+    for {
+      maybeRecord <- connectionRepository.findById(recordId)
+      record <- ZIO.fromOption(maybeRecord).mapError(_ => RecordIdNotFound(recordId))
+      updatedRecord <- updateConnectionProtocolState(
+        recordId,
+        ProtocolState.ConnectionRequestPending,
+        ProtocolState.ConnectionRequestSent
+      )
+    } yield updatedRecord
+}
+```
+
+#### Do not type unexpected errors
+
+Defects from lower layers (typically repository) should not be wrapped in a failure, and error case class declarations
+like [this](https://github.com/hyperledger-labs/open-enterprise-agent/blob/b579fd86ab96db711425f511154e74be75583896/connect/lib/core/src/main/scala/io/iohk/atala/connect/core/model/error/ConnectionServiceError.scala#L8)
+should be prohibited.
+
+Also, considering that failures are expected errors from which user can potentially recover, error case class
```

Comment:
```suggestion
Also, considering that expected failure errors from which the user can potentially recover, error case class
```

Reply:
The Initial meaning is also lost here. The idea is => _failures_ are considered _expected_ errors, while _defects_ are considered _unexpected_ errors.